modules/core/src/com/haulmont/cuba/core/app/Emailer.java

```diff
@@ -138,23 +139,37 @@ protected void prepareEmailInfo(EmailInfo emailInfo) {
         }
     }
 
+    protected void processCaptionTemplate(EmailInfo emailInfo) {
+        String caption = buildStringWithTemplateParams(emailInfo, emailInfo.getCaption());
+        emailInfo.setCaption(caption);
+    }
+
     protected void processBodyTemplate(EmailInfo info) {
         String templatePath = info.getTemplatePath();
-        if (templatePath == null) {
-            return;
+        String templateContents;
+        if (templatePath != null) {
+            templateContents = resources.getResourceAsString(templatePath);
+            if (templateContents == null) {
+                throw new IllegalArgumentException("Could not find template by path: " + templatePath);
+            }
+        }else{
+            templateContents = info.getBody();
         }
 
+        templateContents = buildStringWithTemplateParams(info, templateContents);
```

Comment:
And here too. Do not try to call template processor if params are empty, this will break existing code

Reply:
In this case, `buildStringWithTemplateParams` initialize the map if its null. isn't that ok?

Reply:
You still call > TemplateHelper.processTemplate(templateContents, params);
even for empty maps.

FreeMarker will process it even if Map is empty. It is not OK. Body/subject can contain symbols that will be interpreted as FreeMarker commands.

Reply:
so, if not null instead?
Just to make sure I'm not missing something here. The original code does also that: https://github.com/cuba-platform/cuba/blob/e7098974c9eca1c3e3d61e00cf205794f284586b/modules/core/src/com/haulmont/cuba/core/app/Emailer.java#L147

Reply:
Now, processBodyTemplate is not performed for body and subject if templatePath is null.

Please pass `<#assign link = "x">` as subject to your code without template params (empty template params). It should be inserted to email subject as is, but your code performs processing in this case.

Reply:
This breaks compatibility with existing code where developers do not pass parameters